Attempt: QuestionRenderer mit iFrames

tumidi opened the pull request:
**Hängt ab von #178**

Rendert Attempt innerhalb eines iframes im Vue Frontend (#157).

Da der PR recht umfangreich ist, hier noch ein paar Anmerkungen/Erklärungen:

### Frontend

- `DisplayOptions` werden komplett clientseitig gemanagt
  - vorher wurden die immer zum server gesendet und in browser cookies gespeichert
  - jetzt kümmert sich darum ein pinia store (mit LocalStorage persistence)
  - Display Options werden beim Abrufen des Attempts als query params mitgesendet

### Webserver

- Webserver hat jetzt
  - Controller/View-Architektur, d.h.
    - Trennung von Request/response handling (View) und "Business logic" (Controller)
    - View (Request handling, query params, json-(De-)serialisierung, etc.)
      - Views sind jetzt class-based
    - Controller haben keinen Zugriff auf Request-Objekt
      - Werden durch eine Middleware den View-Objekten automatisch bereitgestellt und haben Zugriff auf "Services", wie `StateManager`, `WorkerPool` etc.
    - aiohttp-jinja2 wird nicht mehr benutzt (sondern jinja2 direkt)
      - Hintergrund: aiohttp-jinja2 verlangt den Request durchzuschleifen (die im Controller aber nicht mehr vorliegen). aiohttp-jinja2 liest intern ein optionales Context-Objekt vom Request, ein Feature, was wir aber sowieso nicht nutzen.
      - `jinja2.ext.do` Extension entfernt: Konnte nicht finden, dass das überhaupt genutzt wird, daher entfernt.
  - Middleswares, die View-übergreifende Aufgaben übernehmen
- Error handling
  - Middlewares, die Application errors auf HTTPExceptions mappen und für die API den Traceback serialisiert, so dass er im Frontend direkt angezeigt wird (macht, denke ich, Sinn für ein SDK)
    - Serialisiert Pydantics `ValidationError`s
  - UIRenderer Errors: Werden serialisiert und im Frontend angezeigt
    - Hier müsste `questionpy_sdk.webserver.controllers.attempt.errors` evtl. noch mal angefasst werden.
    - Aktuell werden die error strings ohne Formatierung im Frontend ausgegeben.
    - Die aktuelle Implementierung baut HTML-Strings (mit `<code>`-Elementen). Um die Formatierung auch im neuen Frontend zu erhalten müsste man die `html_message`-Property mit `to_json` o.ä. ersetzen und Text-Nodes schicken, die dann im Frontend mit den korrekten Tags gerendert werden.
- `StateManager`
  - aus `WebServer` herausgelöst
  - Funktionalität identisch zu vorher, aber soll als Basis dienen, demnächst #165 umzusetzen.

tumidi committed:
feat(frontend): render attempt in iframe

Ref: #157

tumidi committed:
feat: add attempt API routes

Introduce routes for handling attempts in frontend. Some major refactoring
of the webserver code.

Separate business logic (controllers) from request handling (views). Introduce
error middleware. Add StateManager as a starting point for implementing #165.

Ref: #157

tumidi committed:
test: update current webserver implementation tests

Ref: #162

janbritz commented:
Ich habe nicht geschaut, ob das schon vor diesem PR der Fall war:

Wenn ich mir die Question-Options anschaue (`/question`) und dann für ein paar Sekunden den Tab wechsel, werden beim Zurückwechseln meistens zwei GET-Requests gesendet (nach `/options` und `/state`) und nach längerer Zeit auch Drei (`/attempt?...`).

Ähnliches passiert auch bei `/` und `/attempt`, dort werden GET-Requests an `/manifest`, `/options`, `/state` und manchmal `/attempt?...` gesendet. Stammt das von Vue? Aktuell führt das dazu, dass die Logs gespammt werden.

tumidi commented:
> Wenn ich mir die Question-Options anschaue (`/question`) und dann für ein paar Sekunden den Tab wechsel, werden beim Zurückwechseln meistens zwei GET-Requests gesendet (nach `/options` und `/state`) und nach längerer Zeit auch Drei (`/attempt?...`).
> 
> Ähnliches passiert auch bei `/` und `/attempt`, dort werden GET-Requests an `/manifest`, `/options`, `/state` und manchmal `/attempt?...` gesendet. Stammt das von Vue? Aktuell führt das dazu, dass die Logs gespammt werden.

Die Query-Lib hat einen Cache. Es gibt eine Option [`staleTime`](https://pinia-colada.esm.dev/api/@pinia/colada/interfaces/UseQueryOptions.html#staleTime-). Da die Requests ja lokal passieren und near instant sind, habe ich erstmal keine `staleTime` konfiguriert bzw. sehr niedrig gelassen. Man könnte die auch auf 5 Minuten oder noch höher schrauben. Ich denke, das würde potentiell zu stale data führt, wenn z.B. das Paket neu geladen wird, aber die Web app nicht.

Eine andere Variante könnte sein, den `aiohttp.access` logger auf `DEBUG` zu setzen, und den Query-Cache eben gar nicht zu verwenden. Also, wenn es nur um den Log-Spam geht? :thinking: 

```py
access_logger = logging.getLogger('aiohttp.access')
access_logger.setLevel(logging.DEBUG)
```

MHajoha commented on questionpy_sdk/webserver/routes/base.py:
```diff
@@ -0,0 +1,31 @@
+CT = TypeVar("CT", bound="BaseController")
+
+
+class BaseView(web.View, Generic[CT]):
```
Seit Python 3.12 haben Generics eine eigene Syntax:

```suggestion
class BaseView[CT](web.View):
```

Die Zukunft ist jetzt

tumidi replied:
Die Zukunft lässt im SDK noch auf sich warten, wir müssen mal die aktuelle `ruff_defaults.toml` aus qpy-dev synchronisieren (und ganz nebenbei auch dringend mal ruff updaten).

```toml
# ruff_defaults.toml
target-version = "py311"
```

Die Python-Versionen sind irgendwie eh gerade außer Rand und Band. Server will 3.11, SDK 3.12. Ruff ist im SDK noch auf 3.11, GitHub Actions waren noch auf 3.11-13. (Hatte ich in meinem [letzten Commit](https://github.com/questionpy-org/questionpy-sdk/commit/eea89967913c3953e0b36e89a44bb13791dc2446#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR16) angeglichen zum SDK)

Wollten wir nicht mal die letzten 3 Python-Versionen supporten oder so ähnlich? Ich bin verwirrt.

MHajoha replied:
Hm, ja, ich habe bei https://github.com/questionpy-org/questionpy-server/pull/145 ehrlicherweise nicht so ganz an Debian gedacht, aber tatsächlich enthält Debian Bookworm nur 3.11, ich glaube @MartinGauk hatte mal vorgeschlagen, davon wir uns daran orientieren sollten. Wahrscheinlich kommt Debian Trixie noch vor dem nächsten Semester raus und enthält 3.13, aber ich kann den Bump auch zurückrollen, was meint ihr?

Sicherlich macht es aber Sinn, das außerhalb dieses PRs zu betrachten ^^

MHajoha commented on questionpy_sdk/webserver/routes/frontend.py:
```diff
@@ -2,40 +2,28 @@
-        # Catch-all route for all GET requests
-        app.router.add_get("/{path:.*}", handle_spa)
+    # Serve index.html for all other paths
+    return web.FileResponse(STATIC_DIR / "index.html")
```
Wie unterscheidet sich das von [`RouteTableDef.static`](https://docs.aiohttp.org/en/v3.8.1/web_reference.html#aiohttp.web.RouteTableDef.static)?

tumidi replied:
Der Unterschied ist, dass die Route für alle nicht gefundenen Dateien, statt ein 404 zu schicken, die `index.html` ausliefert. Ein Requirement für den client-seitigen Router in der Vue-App. Ansonsten wäre es nicht möglich, direkt "tiefe" Links, wie `/attempt` aufzurufen.

Ich habe versucht von aiohttps `StaticResource` abzuleiten, um den Directory traversal check von dort wiederzuverwenden. Ich denke, das ist nicht wirklich vorgesehen und basiert auf dem Überschreiben von privaten Methoden. Denke diese Lösung ist wesentlich knapper und besser, zumal sie `FileResponse` nutzt, wo eh die meiste relevante Logik drin steckt.

MHajoha commented on frontend/src/queries/fetch.ts:
```diff
@@ -18,19 +19,45 @@ class FetchError extends Error {
+     * @param details An optional detailed description of the error.
      */
-    constructor(status: number, statusText: string, message: string) {
+    constructor(status: number, statusText: string, message: string, details: string | undefined = undefined) {
```
```suggestion
    constructor(status: number, statusText: string, message: string, details?: string) {
```

MHajoha commented on frontend/src/components/common/ErrorDisplay.vue:
```diff
@@ -0,0 +1,33 @@
+  (c) Technische Universität Berlin, innoCampus <info@isis.tu-berlin.de>
+-->
+
+<template>
+    <div v-if="error instanceof FetchError" class="font-monospace">
+        <p>{{ error.message }} ({{ error.status }}: {{ error.statusText }})</p>
+        <code v-if="error.details">
+            <pre>{{ error.details }}</pre>
+        </code>
+    </div>
+    <div v-else-if="error instanceof z.ZodError">
+        <BListGroup>
+            <BListGroupItem v-for="(issue, index) in error.issues" :key="index" class="font-monospace">
+                {{ fromZodIssue(issue) }}
+            </BListGroupItem>
+        </BListGroup>
+    </div>
+    <div v-else class="font-monospace">
+        {{ error.message }}
+    </div>
+</template>
```
Mir fällt gerade auf, dass im Falle von Fehlern nirgends der (JavaScript-)Stacktrace zu finden ist. Den sollten wir aber denke ich schon zumindest loggen.

tumidi replied:
Die Komponente `ErrorDisplay` zeigt aktuell nur backend-seitige Fehler im Client an (bzw. Validierungsfehler von Backend-Daten). Den JS-Stacktrace anzuzeigen hätte nicht so viel Mehrwert, da die Fehler in der Fetch-Logik geworfen und ja bereits gracefully gehandlet und angezeigt wurden.

Ansosten werden clientseitige Fehler nach wie vor durch Vue gehandlet und die Fehler auch in der Browser console geloggt.

Der dritte else-Zweig sollte nicht auftreten. Falls doch, ist das ein Bug im Client. Evtl. wäre es besser diesen Fall zu entfernen und stattdessen irgendwie sowas in der Komponente zu machen?

```
watch(
  () => props.error,
  (error) => {
    if (!(error instanceof FetchError || error instanceof z.ZodError)) {
      throw error;
    }
  },
  { immediate: true }
)
```

MHajoha replied:
Ich hatte aufgrund des erwähnten `optionsFormDataValueSchema akzeptiert weniger (strings & booleans), als Pakete generieren dürfen (zusätzlich int, float, None)`-Punktes einen Zod-Error, da wäre ein Stacktrace hilfreich gewesen, um schneller das Schema zu finden.

Zu dem Vorschlag: Ich nehme an, Pinia loggt einfach nur, wenn Watcher schmeißen? Dazu steht nichts in der Doku. Ich würde eher Vorschlagen, den Fehler dort zu loggen, wo er geworfen wird, also in `get` und `post`, und in den Catch-Blöcken

tumidi replied:
Ich denke, das Problem ist dann, dass `z.Error` nicht den Namen des Zod-Schemas beinhaltet, daher konnte ich nur die Validation-Details anzeigen. Das Schema hat auch keinen "Namen" so gesehen, sondern nur einen JS-Bezeichner, der ja in Production potentiell auch gemangled ist...

Wie dem auch sei, ich logge jetzt die Stack traces im Falle von Validation errors: 65137cdd9421c389936ad30f6dd489e5d44132b6

MHajoha commented on questionpy_sdk/webserver/controllers/base.py:
```diff
@@ -0,0 +1,86 @@
+        except FileNotFoundError as err:
+            if allow_missing:
+                return {}
+            raise web.HTTPConflict(text="Last attempt data not found") from err
```
Ist der Kram nicht besser im `AttemptController` aufgehoben?

MHajoha commented on questionpy_sdk/webserver/controllers/base.py:
```diff
@@ -0,0 +1,86 @@
+        except FileNotFoundError:
+            return None
+
+    async def _get_last_attempt_data(self, *, allow_missing: bool = False) -> Any:
+        try:
+            return await self._state_manager.read_last_attempt_data()
+        except FileNotFoundError as err:
+            if allow_missing:
+                return {}
+            raise web.HTTPConflict(text="Last attempt data not found") from err
```
1. `Any` sollte weitestgehend vermieden werden. Wenn es wirklich alles sein kann, ist `object` besser, hier können wir es doch aber auf ein `dict[str, JsonValue] | None` einschränken, oder?
2. Dass noch keine Response submitted wurde muss anders behandelt werden als der Fall, dass eine leere Response submitted wurde.
3. Warum `HTTPConflict`? Das ist schon eine ungewöhnlich breite Auffassung von "Konflikt"

tumidi replied:
> 1. `Any` sollte weitestgehend vermieden werden. Wenn es wirklich alles sein kann, ist `object` besser, hier können wir es doch aber auf ein `dict[str, JsonValue] | None` einschränken, oder?

Ja, `dict[str, JsonValue]` is auf jeden Fall besser.

> 2. Dass noch keine Response submitted wurde muss anders behandelt werden als der Fall, dass eine leere Response submitted wurde.

Inwiefern? Habe mich an den alten Code gehalten. Dort heißt es:

```py
last_attempt_data = json.loads(webserver.read_state_file(StateFilename.LAST_ATTEMPT_DATA) or "{}")
```

Also, `read_state_file` gibt `None` zurück, falls die Datei nicht existiert, welches dann auf `"{}"` gemappt wird. Es werden beide Fälle gleich behandelt (keine Response/leere Response).

> 3. Warum `HTTPConflict`? Das ist schon eine ungewöhnlich breite Auffassung von "Konflikt"

Ja, 409 ist in dem Fall nicht so toll. 409 habe ich an einigen Stellen benutzt, also immer dort, wo quasi eine "Dependency" fehlt, um den Request zu bedienen. In dem Fall, wenn score aufgerufen wird, aber keine attempt data da ist. Ich denke, wahrscheinlich am ehesten 400, oder?

Wahrscheinlicher ist es auch cleaner, dort überhaupt keine aiohttp-Errors zu schmeißen, sondern eine custom exception zu definieren und die dann im Request handler zu mappen.

MHajoha replied:
> > 2. Dass noch keine Response submitted wurde muss anders behandelt werden als der Fall, dass eine leere Response submitted wurde.
> 
> Inwiefern? Habe mich an den alten Code gehalten. Dort heißt es:
> 
> ```python
> last_attempt_data = json.loads(webserver.read_state_file(StateFilename.LAST_ATTEMPT_DATA) or "{}")
> ```
> 
> Also, `read_state_file` gibt `None` zurück, falls die Datei nicht existiert, welches dann auf `"{}"` gemappt wird. Es werden beide Fälle gleich behandelt (keine Response/leere Response).

Yikes, dann war es auch vorher falsch, sorry. Ich würde es trotzdem hier korrekt machen. `None` heißt "keine Response submitted", `{}` heißt "leere Response submitted"

> > 3. Warum `HTTPConflict`? Das ist schon eine ungewöhnlich breite Auffassung von "Konflikt"
> 
> Ja, 409 ist in dem Fall nicht so toll. 409 habe ich an einigen Stellen benutzt, also immer dort, wo quasi eine "Dependency" fehlt, um den Request zu bedienen. In dem Fall, wenn score aufgerufen wird, aber keine attempt data da ist. Ich denke, wahrscheinlich am ehesten 400, oder?
> 
> Wahrscheinlicher ist es auch cleaner, dort überhaupt keine aiohttp-Errors zu schmeißen, sondern eine custom exception zu definieren und die dann im Request handler zu mappen.

400 passt :+1: 
Strenggenommen wäre es cleaner, für mich ist das aber so fein

tumidi replied:
> Yikes, dann war es auch vorher falsch, sorry. Ich würde es trotzdem hier korrekt machen. `None` heißt "keine Response submitted", `{}` heißt "leere Response submitted"

Fix in 7a522685e481be6fcef2cde34241cc4be55df62d

> 400 passt 👍 Strenggenommen wäre es cleaner, für mich ist das aber so fein

Fix in b3c66e915be9621752fe4bc018b20c0a84409faf

MHajoha commented on questionpy_sdk/webserver/middlewares/error.py:
```diff
@@ -0,0 +1,55 @@
+
+def format_error(err: Exception) -> str:
+    tb = getattr(err, "__traceback__", sys.exc_info()[2])
+    return "".join(traceback.format_exception(type(err), err, tb))
```
Würde hier nicht `traceback.format_exception(err)` reichen? (Dass `format_exception` auch ein Exception-Objekt nimmt, ist nicht auf den ersten Blick ersichtlich, aber der Fall.)

MHajoha commented on questionpy_sdk/webserver/middlewares/error.py:
```diff
@@ -0,0 +1,55 @@
+        raise
+    except Exception as err:
+        traceback.print_exc()
+        raise HTTPInternalServerError(text=f"Server error:\n{format_error(err)}") from err
```
Statt `print_exc` lieber `Logger.exception`

MHajoha reviewed (changes requested):
Mir fallen gerade noch zwei Probleme auf, die nicht durch diesen PR kommen:
- Fragen mit leeren Options-Forms funktionieren nicht, da die "Save"-Knöpfe nie enabled werden.
- `optionsFormDataValueSchema` akzeptiert weniger (strings & booleans), als Pakete generieren dürfen (zusätzlich `int`, `float`, `None`)

tumidi committed:
fix(frontend): only display `FetchError`/`ZodError` in `ErrorDisplay`

Other error types are thrown as they indiciate a client bug.
Log stack trace to browser console in case of a validation error.

tumidi committed:
chore: update TypeScript constructor parameter

tumidi committed:
chore: move attempt-related methods to `AttemptController`

tumidi committed:
chore: use `dict[str, JsonValue]` as type for attempt data

tumidi committed:
refactor: raise aiohttp errors only in views, add custom `MissingStateError`s

tumidi committed:
chore: improve error middleware exception logging and formatting

tumidi committed:
test(frontend): add remaining `formDataUtils` tests

tumidi committed:
chore(frontend): add default query options to mitigate frequent API queries

tumidi committed:
refactor(frontend): implement `onSuccess` callback in `usePostOptionsFormData` to handle attempt restart

tumidi committed:
fix: handle empty dict in `_convert_repetition_dict_to_list`

Add a basic test for empty input.

tumidi committed:
feat(frontend): support option forms with no editable elements

Option forms may contain no editable elements or be empty. Update display logic to
handle this scenario appropriately.

tumidi committed:
fix: distinguish between no attempt (`None`) and empty attempt data (`{}`)

Make `AttemptController._get_last_attempt_data()` return None for missing attempts.

tumidi committed:
refactor(frontend): rename file `useManifest.ts` -> `manifest.ts`

tumidi committed:
fix(frontend): broaden allowed types for `OptionsFormData`

tumidi commented:
> Wenn ich mir die Question-Options anschaue (`/question`) und dann für ein paar Sekunden den Tab wechsel, werden beim Zurückwechseln meistens zwei GET-Requests gesendet (nach `/options` und `/state`) und nach längerer Zeit auch Drei (`/attempt?...`).
> 
> Ähnliches passiert auch bei `/` und `/attempt`, dort werden GET-Requests an `/manifest`, `/options`, `/state` und manchmal `/attempt?...` gesendet. Stammt das von Vue? Aktuell führt das dazu, dass die Logs gespammt werden.

839cfdfb2877b3facbadcb1e103b9dd0a4da0f7b setzt jetzt default fetch options, die für unseren use case, glaube ich, besser sind.

tumidi commented:
>     * Fragen mit leeren Options-Forms funktionieren nicht, da die "Save"-Knöpfe nie enabled werden.

Fix in 85f9660f25f7490dc8714c0072a7ba4be39f125f

>     * `optionsFormDataValueSchema` akzeptiert weniger (strings & booleans), als Pakete generieren dürfen (zusätzlich `int`, `float`, `None`)

Fix in b8915d0fa7ea8f789007152fb6be1babcb5cf562
